Validate design documents.

nickva opened the pull request:
Jira: COUCHDB-2818

Validates:

  * options,filters,lists,shows,updates,views values at
    top level of document are objects

  * validate_doc_upate,language are strings

  * rewrites is an array

  * each view inside the views object is an object

  * each view object has a map function that is a string

  * if a view object has a reduce function it is a string

On validation error it returns a 400 error code with exact
reason what field has failed to validate.

Example:

```
http PUT $DB1/db2/_design/des1 views:='{ "m":"bad"  }'{"ok":true}
HTTP/1.1 400 Bad Request
{
    "error": "invalid_design_doc",
    "reason": "View m must be an object"
}
```

kxepal commented:
1. Need tests
2. Need also check show/list/update/filter functions, views.lib and options
3. May be instead of having a lot of validate_* function just have one `validate_field(Field, ...)` like `validate(<<"view">>, Props)` that dispatches validation routines per field basis?
4. Could you please put space after comma?

But idea is good!

janl commented:
I like this! +1

nickva commented:
I refactored code as suggested. 

1. Now have 30 some tests that should cover most validation code paths.

2. Added validation of shows, lists, updates, filters functions. Special cases for views.lib (looks like a view but it is not one), and options (not a mapping of name : contents, so I let it take any values for now).  

3. I tried to refactor validate_* function to re-use as much as possible. They ended up being per-type (validate_object, validate_array, validate_string).

4. Spaces after commas.

Kxepal, janl, eiri thanks for feedback and for help in #couchdb-dev.

kxepal commented:
@nickva 

> I tried to refactor validate_* function to re-use as much as possible. They ended up being per-type (validate_object, validate_array, validate_string).

Basically:

```
diff --git a/src/couch_mrview.erl b/src/couch_mrview.erl
index 819f580..0472f2e 100644
--- a/src/couch_mrview.erl
+++ b/src/couch_mrview.erl
@@ -46,16 +46,85 @@
 }).
 
 
-validate(DbName, DDoc) ->
-    {Fields} = DDoc#doc.body,
-    case couch_util:get_value(<<"options">>, Fields, {[]}) of
-        {_} -> ok;
-        _ -> throw({invalid_design_doc, <<"`options` parameter must be an object.">>})
-    end,
-    case couch_util:get_value(<<"views">>, Fields, {[]}) of
-        {_} -> ok;
-        _ -> throw({invalid_design_doc, <<"`views` parameter must be an object.">>})
-    end,
+
+validate_ddoc_fields(DDoc) ->
+    lists:foreach(fun(Path) ->
+        validate_ddoc_fields(DDoc, Path)
+    end, [
+        [{<<"filters">>, object}, {any, string}],
+        [{<<"filters">>, object}, {any, string}],
+        [{<<"filters">>, object}, {any, string}],
+        [{<<"language">>, string}],
+        [{<<"lists">>, object}, {any, string}],
+        [{<<"options">>, object}],
+        [{<<"rewrites">>, array}],
+        [{<<"shows">>, object}, {any, string}],
+        [{<<"updates">>, object}, {any, string}],
+        [{<<"validate_doc_update">>, string}],
+        [{<<"views">>, object}, {any, object}, {<<"map">>, string}],
+        [{<<"views">>, object}, {any, object}, {<<"reduce">>, string}],
+        [{<<"views">>, object}, {<<"lib">>, object}, {any, string}]
+    ]),
+    ok.
+
+validate_ddoc_field(undefined, Type) when is_atom(Type) ->
+    ok;
+validate_ddoc_field(_, any) ->
+    ok;
+validate_ddoc_field(Value, string) when is_binary(Value) ->
+    ok;
+validate_ddoc_field(Value, array) when is_list(Value) ->
+    ok;
+validate_ddoc_field({Value}, object) when is_list(Value) ->
+    ok;
+validate_ddoc_field({Props}, {any, Type}) ->
+    lists:foldl(fun
+        ({_, Value}, ok) ->
+            validate_ddoc_field(Value, Type);
+        ({Key, _}, error) ->
+            {error, Key}
+    end, ok, Props);
+validate_ddoc_field({Props}, {Key, Type}) ->
+    validate_ddoc_field(couch_util:get_value(Key, Props), Type);
+validate_ddoc_field(_, _) ->
+    error.
+
+validate_ddoc_fields(DDoc, Path) ->
+    case validate_ddoc_fields(DDoc, Path, []) of
+        ok -> ok;
+        {error, {FailedPath0, Type0}} ->
+            FailedPath = iolist_to_binary(lists:reverse(FailedPath0)),
+            Type = ?l2b(atom_to_list(Type0)),
+            throw({invalid_design_doc,
+                  <<"`", FailedPath/binary, "` must be ", Type/binary>>})
+    end.
+
+validate_ddoc_fields(undefined, _, _) ->
+    ok;
+validate_ddoc_fields(_, [], _) ->
+    ok;
+validate_ddoc_fields({KVS}=Props, [{any, Type} | Rest], Acc) ->
+    lists:foldl(fun
+        ({Key, _}, ok) ->
+            validate_ddoc_fields(Props, [{Key, Type} | Rest], Acc);
+        ({_, _}, {error, _}=Error) ->
+            Error
+    end, ok, KVS);
+validate_ddoc_fields({KVS}=Props, [{Key, Type} | Rest], Acc) ->
+    case validate_ddoc_field(Props, {Key, Type}) of
+        ok ->
+            validate_ddoc_fields(couch_util:get_value(Key, KVS),
+                                 Rest,
+                                 [Key | Acc]);
+        error ->
+            {error, {[Key, <<".">> | Acc], Type}};
+        {error, Key1} ->
+            {error, {[Key1, <<".">> | Acc], Type}}
+    end.
+
+
+validate(DbName,  DDoc) ->
+    ok = validate_ddoc_fields(DDoc#doc.body),
     GetName = fun
         (#mrview{map_names = [Name | _]}) -> Name;
         (#mrview{reduce_funs = [{Name, _} | _]}) -> Name;

```

If you'll not match exact error message in test you'll see that all will pass. Except requirement for map function - that's not true case since you may have just `views.lib` commonjs module with no any map/reduce function around due to legacy reasons or with intention to add views later and reuse the same code.

nickva commented:
@kxepal Thanks for the review!

I updated to not validate the contents of views.lib object, so lib can contain anything now. In the code see "validate_view/2 (first clause) and unit test should_accept_any_in_lib/1 tests that.

kxepal commented:
@nickva cool! Do you want accept my patch? I think it's a bit more Erlang`ish (: However, it need to be improved in few places as I didn't spend much time for make it well.

Also, issue with map-less views still need to be addressed. If you need some real case: geocouch. While it uses `spatial` object for index functions, it CommonJS modules picks from `views.lib`, so you may have a design document like this one (picked from geocouch tests):
```
{
    "_id": "_design/spatial",
    "language": "javascript",
    "views": {
        "lib": {
            "geo": "exports.type = 'Point';"
        }
    },
    "spatial": {
        "basicIndex": "function(doc) {var lib = require('views/lib/geo'); emit({type: lib.type, coordinates: [doc.loc[0], doc.loc[1]]}, doc.string);}"
    }
}
```

And while there is everything ok with it, eventually we not allow such ddocs without strong reason why.

nickva commented:
@kxepal  Thanks again for help!

After discussion on #couchdb-dev, allow views without map functions.

kxepal commented:
@nickva thanks for update. Looks like we solved all possible BC issues (except cases when people has already broken design documents).

@rnewson , what would you say?

rnewson commented:
@nickva @kxepal the map function is mandatory, please restore the check. The "spatial" example is _outside_ of the special "views" object and we have no business validating it.

rnewson commented on src/couch_mrview.erl:
```diff
@@ -46,16 +46,103 @@
+        MapVal when is_binary(MapVal) ->
+            ok;
+        undefined ->
+            ok;
```
not ok.

rnewson commented on src/couch_mrview.erl:
```diff
@@ -46,16 +46,103 @@
+%%    "lst2" : "function(head, req){...}"
+%% }
+%%
+%% With 2 excpetions:
```
typo

rnewson commented on src/couch_mrview.erl:
```diff
@@ -46,16 +46,103 @@
+
+%% Validate top level design document objects.
+%%
+%% Most cases (shows,lists,fitlers,...) will be
```
typo

rnewson commented on src/couch_mrview.erl:
```diff
@@ -46,16 +46,103 @@
+%%  - views :  Validated by a special function.
+%%  - options : Allowed to contain non-string values.
+%%
+validate_opt_object(<<"views">>, {Views})->
```
inferring `opt` is short for `optional`, but `optional` would mean no one had to.

rnewson commented on src/couch_mrview.erl:
```diff
@@ -46,16 +46,103 @@
+
+
+%% If this field value is present, it must be a string
+validate_opt_string(FName, FVal)->
```
spacing

rnewson commented on test/couch_mrview_ddoc_validation_tests.erl:
```diff
@@ -82,3 +109,254 @@ should_reject_invalid_builtin_reduce(Db) ->
+                  <<"View view1 must be an object">>},
+                  couch_db:update_doc(Db, Doc, [])).
+
+should_accept_view_without_map_function(Db) ->
```
no it shouldn't :)

rnewson commented:
+1 (cc @kxepal any further comments?)

kxepal commented:
No.
